pkg/controller/factory/base_controller.go

```diff
@@ -0,0 +1,158 @@
+	terminationGraceDuration time.Duration
+}
+
+var _ Controller = &baseController{}
+
+func (c *baseController) Run(ctx context.Context, workers int) {
+	// shutdownContext is used to signal whether a controller has gracefully shutdown
+	shutdownContext, shutdownComplete := context.WithCancel(context.TODO())
+	if c.shutdownContext != nil {
+		panic(fmt.Sprintf("controller %q is already running", c.name))
+	}
+	c.shutdownContext = shutdownContext
+
+	// HandleCrash recovers panics
+	defer utilruntime.HandleCrash()
+
+	klog.Infof("Starting %s ...", c.name)
+	if !cache.WaitForCacheSync(ctx.Done(), c.cachesToSync...) {
+		// this will be recovered using HandleCrash()
+		panic("timeout waiting for informer cache")
+	}
+	klog.V(5).Infof("Caches synced for controller %s", c.name)
+
+	var workerWaitGroup sync.WaitGroup
+
+	// queueContext is used to track and initiate queue shutdown
+	queueContext, queueShutdownComplete := context.WithCancel(context.TODO())
+
+	for i := 1; i <= workers; i++ {
+		klog.Infof("Starting #%d worker of %s controller ...", i, c.name)
+		workerWaitGroup.Add(1)
+		go func() {
+			defer func() {
+				klog.Infof("Shutting down worker of %s controller ...", c.name)
+				workerWaitGroup.Done()
+			}()
+			c.runWorker(queueContext)
+		}()
+	}
+
+	// runPeriodicalResync is independent from queue:w
+	go c.runPeriodicalResync(ctx, c.resyncEvery)
+
+	// Handle controller shutdown
+
+	<-ctx.Done() // wait for controller context to be cancelled
+	klog.Infof("Shutting down %s ...", c.name)
+
+	c.syncContext.Queue().ShutDown() // shutdown the controller queue first
+	for !c.syncContext.Queue().ShuttingDown() {
+		time.Sleep(10 * time.Millisecond) // don't race too much with the queue
+	}
+	queueShutdownComplete() // signal workers the queue is terminated
+
+	// wait for all workers to finish their jobs
+	workerWaitGroup.Wait()
+
+	shutdownComplete() // cancel the shutdown context
+}
+
+func (c *baseController) Sync(ctx context.Context, syncCtx SyncContext) error {
+	return c.sync(ctx, syncCtx)
+}
+
+func (c *baseController) ShutdownContext() context.Context {
+	return c.shutdownContext
+}
+
+func (c *baseController) runPeriodicalResync(ctx context.Context, interval time.Duration) {
+	if interval == 0 {
+		return
+	}
+	go wait.UntilWithContext(ctx, func(ctx context.Context) {
+		if err := c.sync(ctx, c.syncContext.(syncContext).withRuntimeObject(nil)); err != nil {
+			utilruntime.HandleError(fmt.Errorf("periodical resync of controller %s failed: %v", c.name, err))
+		}
+	}, interval)
+}
+
+// runWorker runs a single worker
+// The worker is asked to terminate when the passed context is cancelled and is given terminationGraceDuration time
+// to complete its shutdown.
+func (c *baseController) runWorker(queueCtx context.Context) {
+	// workerContext is used to signal when the processing of the current item is done and the queue is shutdown
+	workerContext, workerDone := context.WithCancel(context.Background())
+
+	go func() {
```

Comment:
leaking a worker loop

Reply:
how is this leaking a worker loop?

Reply:
where do you wait for that lambda to finish? (WaitGroup)